android/engine/src/main/java/org/smartregister/fhircore/engine/cql/LibraryEvaluator.kt

```diff
@@ -0,0 +1,175 @@
+    bundleFactory.addResourcesToBundle(
+      resources,
+      BundleTypeEnum.COLLECTION,
+      "",
+      BundleInclusionRule.BASED_ON_INCLUDES,
+      null
+    )
+    contentProvider =
+      BundleFhirLibraryContentProvider(
+        fhirContext,
+        bundleFactory.resourceBundle as IBaseBundle,
+        adapterFactory,
+        libraryVersionSelector
+      )
+
+    // Load terminology content, and create a TerminologyProvider which is the interface used by the
+    // evaluator for resolving terminology
+    val valueSetStream: InputStream = ByteArrayInputStream(valueSetData.toByteArray())
+    val valueSetBundle = parser.parseResource(valueSetStream)
+    terminologyProvider = BundleTerminologyProvider(fhirContext, valueSetBundle as IBaseBundle)
+
+    // Load data content, and create a RetrieveProvider which is the interface used for
+    // implementations of CQL retrieves.
+    val dataStream: InputStream = ByteArrayInputStream(testData.toByteArray())
+    val dataBundle = parser.parseResource(dataStream)
+    bundleRetrieveProvider = BundleRetrieveProvider(fhirContext, dataBundle as IBaseBundle)
+    bundleRetrieveProvider!!.terminologyProvider = terminologyProvider
+    bundleRetrieveProvider!!.isExpandValueSets = true
+    cqlEvaluator =
+      CqlEvaluator(
+        object :
+          TranslatingLibraryLoader(
+            ModelManager(),
+            listOf(contentProvider),
+            CqlTranslatorOptions.defaultOptions()
+          ) {},
+        object : HashMap<String?, DataProvider?>() {
+          init {
+            put(
+              "http://hl7.org/fhir",
+              CompositeDataProvider(R4FhirModelResolver(), bundleRetrieveProvider)
+            )
+          }
+        },
+        terminologyProvider
+      )
+    libraryEvaluator = LibraryEvaluator(cqlFhirParametersConverter, cqlEvaluator)
+  }
+
+  /**
+   * This method is used to run a CQL Evaluation
+   * @param libraryData Fhir resource type Library
+   * @param helperData Fhir resource type LibraryHelper
+   * @param valueSetData Fhir resource type ValueSet
+   * @param testData Fhir resource to evaluate e.g Patient
+   * @param evaluatorId Outcome Id of evaluation e.g ANCRecommendationA2
+   * @param context Fhir context, eg. Patient
+   * @param contextLabel Fhir context e.g. mom-with-anemia
+   * @return JSON String Fhir Resource type Parameter
+   */
+  fun runCql(
+    libraryData: String,
+    helperData: String,
+    valueSetData: String,
+    testData: String,
+    evaluatorId: String?,
+    context: String,
+    contextLabel: String
+  ): String {
+    loadConfigs(libraryData, helperData, valueSetData, testData)
+    val result =
+      libraryEvaluator!!.evaluate(
+        VersionedIdentifier().withId(evaluatorId),
+        Pair.of(context, contextLabel),
+        null,
+        null
+      )
+    return fhirContext.newJsonParser().encodeResourceToString(result)
```

Comment:
Do we want to always return a string after running the evaluation or do we maybe return a resource and leave it up to the user to convert if they need to?

Alternatively, we can overload this method and have one that returns the _Resource result_ and one that returns the _String result._

Reply:
@ndegwamartin. Having this class only have only one responsibility is what I intended. This will allow other members to focus on other logic and not worry about parsing and loading fhir resource. The string returned is a JSON which will be used to display results. We don't intend to use the String response on other logic.  This will also help in Unit tests.

Reply:
Cool, if the CQL evaluation results will always only ever be used as as string this approach is ok. 

I'm marking this as resolved